### PR TITLE
Fix deadlocks

### PR DESCRIPTION
**Pull request by nyalldawson**
Aimed at fixing #43572 (although the script referenced in that issue also needs fixes, which I'll comment in that ticket).

This PR adds a QgsFeedback argument to QgsFeatureRequest, so that we can abort feature iteration early in the situation that the results are no longer required (e.g. when a layer renderer is canceled).

While we already have QgsFeatureIterator::setInterruptionChecker which is designed to handle a similar situation, we can't use that particular API to fix the deadlock from #43572. Specifically, we need the feedback set on the request so that we can handle cancelation requests which occur during the construction of QgsFeatureIterators, rather than only those which occur after the iterator is created. In this particular situation the deadlock happens while acquiring a connection to the ogr layer, so I've add a  QgsFeedback argument to QgsConnectionPool::acquireConnection -- if set, this causes the acquisition to be attempted every 300 ms until either the connection is acquired, the feedback is canceled, OR the overall timeout length is reached.

That partially helps #43572, but to solve #43572 properly we also need a way to cancel unwanted expression evaluation for costly expressions (such as the custom expression function attached to that ticket). To handle this I've introduced a QgsFeedback object to QgsExpressionContext also, so that potentially expensive expression functions have a means to test if the result is no longer wanted and abort the evaluation early. (I've added checks here for some potentially expensive functions, specifically those which directly iterate features during evaluation. There's more scope to add this optimisation in other expression functions, e.g. ~~the aggregate functions~~ done!). 

With these two changes combined we now have a proper means to abort the vector layer feature counter early and avoid the unkillable tasks described in #43572.

This fix is too risky to merge for 3.20, so targeting it for >= 3.22 only.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@elpaso @rouault would appreciate your eyes on this one!

File: src/core/vector/qgsvectorlayerfeaturecounter.cpp [hunk 1]
```diff
@@ -91,16 +97,26 @@ bool QgsVectorLayerFeatureCounter::run()
       if ( isCanceled() )
       {
         mRenderer->stopRender( renderContext );
+        mFeedback.reset();
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayerfeaturecounter.cpp:100]: does this need the feedback reset on the expression context too here?

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerfeaturecounter.cpp:100]: Good catch, yes

**Comment by m-kuhn** [comment]
Thanks for looking deeper, a good step in the right direction :+1:

File: src/core/qgsaggregatecalculator.h [hunk 2]
```diff
@@ -171,10 +171,11 @@ class CORE_EXPORT QgsAggregateCalculator
      * If an expression is used, then the context parameter must be set.
      * \param context expression context for evaluating expressions
      * \param ok if specified, will be set to TRUE if aggregate calculation was successful
+     * \param feedback optional feedback argument for early cancellation (since QGIS 3.22)
```
Comment (elpaso) [inline src/core/qgsaggregatecalculator.h:174]: Maybe add a line about this taking precedence over the feedback from the context.

**Review by elpaso — APPROVED** [review]
Nice work, thank you!

File: src/core/vector/qgsvectorlayerfeaturecounter.h [hunk 3]
```diff
@@ -95,8 +97,9 @@ class CORE_EXPORT QgsVectorLayerFeatureCounter : public QgsTask
     QgsExpressionContext mExpressionContext;
     QHash<QString, long> mSymbolFeatureCountMap;
     QHash<QString, QgsFeatureIds> mSymbolFeatureIdMap;
+    std::unique_ptr< QgsFeedback > mFeedback;
     bool mWithFids = false;
-    int mFeatureCount;
+    long mFeatureCount = 0;
```
Comment (rouault) [inline src/core/vector/qgsvectorlayerfeaturecounter.h:102]: For Win64:
```suggestion
    qint64 mFeatureCount = 0;
```

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerfeaturecounter.h:102]: I've matched it to QgsVectorLayer::featureCount()/QgsVectorDataProvider::featureCount(), which all use longs.

Comment (rouault) [reply src/core/vector/qgsvectorlayerfeaturecounter.h:102]: > I've matched it to QgsVectorLayer::featureCount()/QgsVectorDataProvider::featureCount(), which all use longs.

ah, they are wrong then. I'll send a PR to fix them

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerfeaturecounter.h:102]: Can I ask what the issue is with the long type? It's been this way for decades :laughing:

Comment (rouault) [reply src/core/vector/qgsvectorlayerfeaturecounter.h:102]: > Can I ask what the issue is with the long type? It's been this way for decades laughing

That's a common portability issue: long is 32-bit on Windows 64 bit, whereas it is 64-bit on Linux 64 bit : https://stackoverflow.com/questions/384502/what-is-the-bit-size-of-long-on-64-bit-windows

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerfeaturecounter.h:102]: Could we just go "long long" instead then. and avoid the qt type?

Comment (rouault) [reply src/core/vector/qgsvectorlayerfeaturecounter.h:102]: that can do

**Pull request merged**